hnn_core/cell.py

```diff
@@ -137,26 +141,134 @@ def gid(self, gid):
+        # origin by passing cell.soma as a sec argument to h.distance()
+        # Then iterate over segment nodes of dendritic sections
+        # and set gbar_ar depending on h.distance(seg.x), which returns
+        # distance from the soma to this point on the CURRENTLY ACCESSED
+        # SECTION!!!
+        h.distance(sec=self.soma)
+        for sec in self.sections:
+            sec_name = sec.name().split('_', 1)[1]
+            for mech_name, mech in p_secs[sec_name]['mechs'].items():
+                sec.insert(mech_name)
+                for attr, val in mech.items():
+                    if hasattr(val, '__call__'):
+                        sec.push()
+                        for seg in sec:
+                            setattr(seg, attr, val(h.distance(seg.x)))
+                        h.pop_section()
+                    else:
+                        setattr(sec, attr, val)
+
+    def _create_synapses(self, p_secs, p_syn):
+        """Create synapses."""
+        for sec_name in p_secs:
+            for receptor in p_secs[sec_name]['syns']:
+                sec_name_sanitized = sec_name.replace('_', '')
+                syn_key = f'{sec_name_sanitized}_{receptor}'
+                if sec_name == 'soma':
+                    seg = self.soma(0.5)
+                else:
+                    seg = self.dends[sec_name](0.5)
+                self.synapses[syn_key] = self.syn_create(
+                    seg, **p_syn[receptor])
+
+    def _create_sections(self, p_secs):
+        """Create soma and set geometry.
+
+        Notes
+        -----
+        By default neuron uses xy plane
+        for height and xz plane for depth. This is opposite for model as a
+        whole, but convention is followed in this function ease use of gui.
+        """
+        sec_pts, _, _, _, topology = self.secs
+
+        for sec_name in p_secs:
+            if sec_name == 'soma':
+                self.soma = h.Section(cell=self, name=self.name + '_soma')
+                sec = self.soma
+            else:
+                self.dends[sec_name] = h.Section(
+                    name=self.name + '_' + sec_name)
+                sec = self.dends[sec_name]
+
+            h.pt3dclear(sec=sec)
+            for pt in sec_pts[sec_name]:
+                h.pt3dadd(pt[0], pt[1], pt[2], 1, sec=sec)
+            sec.L = p_secs[sec_name]['L']
+            sec.diam = p_secs[sec_name]['diam']
+            sec.Ra = p_secs[sec_name]['Ra']
+            sec.cm = p_secs[sec_name]['cm']
+
+            if sec.L > 100.:
+                sec.nseg = int(sec.L / 50.)
+                # make dend.nseg odd for all sections
+                if not sec.nseg % 2:
+                    sec.nseg += 1
+
+        if topology is None:
+            topology = list()
+
+        # Connects sections of THIS cell together.
+        for connection in topology:
+            # XXX: risky to use self.soma as default. Unfortunately there isn't
```

Comment:
This always confused me, I am curious if it's possible to make such a dictionary.

Reply:
If `connection[0] = 'blah'` then `parent_sec = self.soma` which is not correct. It should have failed instead. I pushed a more readable and safer version in: f2222cc6554c85